### DISCUSSION TITLE
How to get default value of deferred columns for dataclass

### DISCUSSION THREAD

**Discussion by K1DV5** [sqlalchemy/sqlalchemy#10283] [Usage Questions] [answered]
I'm trying to use `asdict` function of dataclasses with `defer()` applied on some of the columns. Since it tries to access the deferred column, `sqlalchemy.exc.MissingGreenlet` is raised. I want it to use the default value of the column instead. How do I do that?

I will provide a code snippet if necessary.

**Comment by zzzeek** [comment]
you would need to eagerly load those columns up front (see https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#using-undefer-to-eagerly-load-deferred-columns), or access them ahead of time using AsyncAttrs (see https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs).  an overview of this area is at https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession

**Reply by K1DV5** [reply]
Let's say the model is:
```python
class Model(Base):
    __tablename__ = 'models'

    id: Mapped[int] = mapped_column()
    name: Mapped[str] = mapped_column(String(30))
    content: Mapped[str] = mapped_column(String, default=None)
```
Now let's say I wanted to convert an instance into a dict using `asdict`, but I only want the `id` and `name` fields. I want to avoid eager loading the content column because it may be large. I can't prevent `asdict` from trying to access the attribute. So on what you said:
1. I want to avoid eager loading.
2. I don't want/need the column value so I don't want to use AsyncAttrs which could fetch the data regardless of whether it's large.

Do I have any other choice other than doing it manually, without `asdict`? Because `asdict` is one of the things that made dataclasses support attractive for me.

**Reply by CaselIT** [reply]
Relationships have the noload strategy, but it's not available for notmal column.

Note that the use case you want is not supported by asdict: you can't tell it to only return q subset of keys

**Reply by K1DV5** [reply]
I know. I was hoping to have the default values (like noload you mentioned) instead of a subset of keys. Thanks anyway.

**Reply by CaselIT** [reply]
While in general it does not make much sense to have noload on a normal column, I'm not sure if there are specific reason for allowing it only on relationships. In the end it does not make much sense there too, since it's just setting some arbitrary data that's not what in the db.

@zzzeek do you have a particular opinion here?

**Reply by zzzeek** [reply, marked as answer]
ive always maintained that getting a serialization like this:

```py
{
   "name": "the name",
   "street": "street address",
   "state": "NY",
   "zip": "12345",
   "description": None
}
```

where above, "description" would be some kind of "noload" element where we didn't want to load the data from the database, is a bad idea.  The dictionary above has incorrect information in it which will lead to bugs and difficult to debug issues.

The dictionary we **do** want is this:

```py
{
   "name": "the name",
   "street": "street address",
   "state": "NY",
   "zip": "12345",
}
```

Where above, the "description" field that we don't need, **is not there**.   That way it isn't a key that's associated with garbage data.

Where this goes is that automatic dicionary makers like `_asdict()` are **not useful for more complicated cases**.   Use regular Python:

```py
mydict = {k: getattr(obj, k) for k in ('name', 'street', 'state', 'zip')}
```

or otherwise use real serialization schemas that represent the different views of data you need, something like [marshmallow](https://marshmallow.readthedocs.io/en/stable/).

**Reply by K1DV5** [reply]
One more question, then how is it justified that there is the `noload` option for relationships? Wouldn't that also be considered incorrect information?

**Reply by zzzeek** [reply]
it was intended as a way for custom loader schemes to populate the attribute instead, driven by events.  there was an example that illustrated this but it's since been removed.    There are people using it in some ways similar to this.

**Comment by zzzeek** [comment]
> Let's say the model is:
> 
> ```python
> class Model(Base):
>     __tablename__ = 'models'
> 
>     id: Mapped[int] = mapped_column()
>     name: Mapped[str] = mapped_column(String(30))
>     content: Mapped[str] = mapped_column(String, default=None)
> ```
> 
> Now let's say I wanted to convert an instance into a dict using `asdict`, but I only want the `id` and `name` fields. 

to do this at scale, use a real serialization library like [marshmallow](https://marshmallow.readthedocs.io/en/stable/quickstart.html#serializing-objects-dumping):

```py
from __future__ import annotations

from marshmallow import fields
from marshmallow import Schema

from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column



class Base(DeclarativeBase):
    pass


class Model(Base):
    __tablename__ = "models"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(30))
    content: Mapped[str] = mapped_column(String, default=None)


m1 = Model(id=1, name="some name", content="some content")

# Now let's say I wanted to convert an instance into a dict using *****,
# but I only want the id and name fields.

# this schema would be set up just once in the application, among all the other schemas that are used 
# for various kinds of serializations that the application needs.
id_and_name = Schema.from_dict({"id": fields.Int(), "name": fields.Str()})()

my_dictionary = id_and_name.dump(m1)

print(my_dictionary)
```

> Do I have any other choice other than doing it manually, 

use a serialization schema as above


> without `asdict`? Because `asdict` is one of the things that made dataclasses support attractive for me.

`asdict()` is the least useful part of dataclasses.   At a basic level, it is fundamentally incompatible with SQLAlchemy models because if you have a bidirectional model, like `User->Address` with a backref from `Address->User`, `asdict()` is now impossible to use because `asdict()` and `astuple()` even moreso by definition can't handle object graphs with cycles in them.

the main things dataclasses bring are 1. generation of complex `__init__()` method signatures 2. generation of methods like `__repr__()`, `__eq__()`. and 3.  ...that's about it, in fact.

**Reply by CaselIT** [reply]
repr is also kind of hit and miss, since it will access all attributes, just like asdict, even if they aren't loaded (I guess eq is the same)

**Reply by zzzeek** [reply]
you can at least put `repr=False` on attributes

**Reply by CaselIT** [reply]
indeed you can

**Reply by zzzeek** [reply]
seems like people prefer pydantic over marshmallow because marshmallow has a reputation for poor performance, which is too bad

**Reply by CaselIT** [reply]
don't know. I've seen multiple test that show pydantic is not that fast compared to other serialization framework (at least v1), but I don't remember how it compares with marshmellow.

I think it became popular in part thanks to fastapi

**Reply by K1DV5** [reply]
@zzzeek You have a point. I'm using fastapi so pydantic is already there. Thank you for the answer.
